#### Aquí encontrarás todo lo que necesitas saber: https://www.kaggle.com/t/ab8726f0cfc84544abbae69a6be88071

## Librerias

In [27]:
import pandas as pd

import urllib.request
from PIL import Image

## Data

In [28]:
### Para que funcione necesitas bajarte los archivos de datos de Kaggle 
train = pd.read_csv("data/train2.csv", index_col=0)


## Data exploration

In [29]:
train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 912 entries, 1104 to 1173
Data columns (total 16 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Company      912 non-null    int64  
 1   TypeName     912 non-null    int64  
 2   Inches       912 non-null    float64
 3   Ram          912 non-null    int64  
 4   OpSys        912 non-null    int64  
 5   Weight       912 non-null    float64
 6   Price_euros  912 non-null    float64
 7   Memory1_1    912 non-null    int64  
 8   Memory1_2    912 non-null    int64  
 9   Memory2_1    912 non-null    int64  
 10  Resolution1  912 non-null    int64  
 11  Gpu1         912 non-null    int64  
 12  Gpu2         912 non-null    int64  
 13  Cpu1         912 non-null    int64  
 14  Cpu2         912 non-null    int64  
 15  Cpu3         912 non-null    int64  
dtypes: float64(3), int64(13)
memory usage: 121.1 KB


In [30]:
train.head()

Company  TypeName  Inches  Ram  OpSys  Weight  Price_euros   
laptop_ID                                                               
1104            19         6    15.6    4      9    2.40        387.0  \
114             20         7    13.3    8      9    1.23       1379.0   
172             19         6    17.3    8      9    3.00        854.0   
918             21         8    12.5    8      9    1.26       1483.0   
447             22         6    17.3    8      9    2.80        519.0   

           Memory1_1  Memory1_2  Memory2_1  Resolution1  Gpu1  Gpu2  Cpu1   
laptop_ID                                                                   
1104             500          4          0      1049088     0     0     0  \
114              256          5          0      5760000     1     1     1   
172              256          5          0      2073600     2     2     1   
918              512          5          0      2073600     1     3     1   
447             1000          4          0      1440000     0     0     0   

           Cpu2  Cpu3  
laptop_ID              
1104         11     0  
114          12     1  
172          12     2  
918          12     1  
447          13     3

## Data processing

-----------------------------------------------------------------------------------------------------------------

## Pasos que si o si debéis realizar para poder participar en la competición:

### 1. Definir X e y

In [31]:
X = train.drop(['Price_euros'], axis=1)
y = train['Price_euros'].copy()
X.shape

(912, 15)

In [32]:
y.shape

(912,)

### 2. Dividir X_train, X_test, y_train, y_test

In [33]:
from sklearn.model_selection import train_test_split

In [34]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 42)

In [35]:
X_train

Company  TypeName  Inches  Ram  OpSys  Weight  Memory1_1   
laptop_ID                                                             
26              20         6    15.6    4      9    2.30       1000  \
776             20         9    15.6   16      9    2.72        128   
1263            23         9    15.6   16      9    2.50        256   
182             22         6    15.6    8      9    1.90        128   
321             21        10    13.3   16      9    1.29       1000   
...            ...       ...     ...  ...    ...     ...        ...   
394             20        11    15.6    8      9    2.00        256   
441             22         6    15.6   16      9    2.50        512   
955             23         6    17.3    4      9    2.80       1000   
210             20         6    15.6    6      9    2.30       1000   
673             28         6    15.6    4      9    2.10        128   

           Memory1_2  Memory2_1  Resolution1  Gpu1  Gpu2  Cpu1  Cpu2  Cpu3  
laptop_ID                                                                   
26                 4          0      1049088     1     3     1    12     4  
776                5       1000        15360     2     2     1    12     1  
1263               5       1000      2073600     2     2     1    12     1  
182                5       1000      2073600     0     8     1    12     2  
321                5          0        15360     1     4     1    12     1  
...              ...        ...          ...   ...   ...   ...   ...   ...  
394                5          0      2073600     2     5     1    12     1  
441                5          0      2073600     2     5     1    12     1  
955                4          0      1440000     2     2     1    14     5  
210                4          0      1049088     1     3     1    12     4  
673                5          0      1049088     1     3     1    12     4  

[729 rows x 15 columns]

### 3. Asignar el modelo (vacío) a una variable


In [36]:
from sklearn.linear_model import LinearRegression

model = LinearRegression()

In [37]:
model.fit(X_train, y_train)

LinearRegression()

In [38]:
prediction = model.predict(X_test)
prediction

array([1059.24153665,   72.9348974 ,  881.51986077, 1407.14850864,
       2082.34838497,  841.65096388, 1058.92769468, 1456.32814943,
        905.99156311, 1186.37897037, 1409.30418518,  473.36633091,
        646.74911034, 1252.46867438, 1506.51182951, 1948.08103685,
       1904.95801249, 1219.33553103,  850.65546848,  437.49350552,
        531.10851719,  629.83768943, 1143.75698368, 1239.97461986,
       1021.50353632,  609.91685163,  912.80035446, 1901.78063959,
       1036.67056848,  379.12304807, 1179.23872167, 1461.74888214,
        763.24707145, 1434.12284935, 1089.47019236,  366.0675553 ,
       1015.98662602,  894.66806587, 1204.64265017,  854.81628181,
       1338.55428453, 1197.23150122,  909.78933937, 1932.19763308,
        948.50087576,  540.18415921, 1614.99325219,  595.17447684,
       1270.11708971,  571.14876359, 1981.25247387,  959.25013987,
        337.67323189, 2179.81303895,  823.47340311, 1225.6722762 ,
       1816.30250497, 1429.15175964, 1883.16299097, 1957.69727

### 4. Sacar métricas, valorar el modelo 

Recuerda que en la competición se va a evaluar con la métrica de MAE.

In [39]:
from sklearn.metrics import mean_squared_error
import numpy as np

print('RMSE :', np.sqrt(mean_squared_error(y_test, prediction)))

RMSE : 454.1072076354975


-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir con el dataset de predicción 

Definición de **modelo que está listo**. 

Tras hacer suficientes pruebas, analizar los datos, hacer feature engineering, probar diferentes modelos con diferentes parámetros, es con este con el que observo mejores métricas y menos overfitting. ¡Cuidado con el overfitting aquí! Si vuestro modelo aprende muy bien de estos datos pero hay overfitting cuando le pasemos los datos desconocidos de `test.csv` nos arriesgamos a que digamos, no salga lo esperado.

### 1. Entrena dicho modelo con TODOS tus datos de train, esto es con `train.csv` al completo.


**CON LAS TRANSFORMACIONES QUE LE HAYAS REALIZADO A `X` INCLUÍDAS.**


Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

### 2. Carga los datos de `test.csv` para predecir.

In [13]:
X_pred = pd.read_csv("data/test2.csv", index_col=0)
X_pred.head()

Company  TypeName  Inches  Ram  OpSys  Weight  Memory1_1   
laptop_ID                                                             
750              9         0    14.0    8      0    1.25        256  \
726             14         0    14.0    4      0    1.40         32   
633             14         0    14.0    2      0    1.40         32   
363              1         3    15.6    8      0    2.65       1000   
319              4         4    11.6    2      0    1.10         32   

           Memory1_2  Memory2_1  Memory2_2  Resolution1  Gpu1  Gpu2  Cpu1   
laptop_ID                                                                   
750                1          0          4      2073600     1     3     1  \
726                0          0          4      2304000     1     3     1   
633                3          0          4      2073600     1     3     1   
363                0          0          4      2073600     2     2     1   
319                3          0          4      1049088     1     3     1   

           Cpu2  Cpu3  
laptop_ID              
750          12     2  
726          21    29  
633          21    15  
363          12     2  
319          15     6

In [14]:
X_pred.shape

(391, 16)

**¿Por qué da error?** 

In [15]:
predictions_submit = model.predict(X_pred)
predictions_submit

array([1226.19776101,   28.35335923,  155.15443347, 1108.91859732,
        728.60547994,  483.48832199,  535.02008634, 1927.19313338,
       1814.89584674, 1314.5536642 , 1084.82894655, 1111.85381163,
       1313.36783648, 1018.88331038,  690.45177683, 1069.20563923,
       1233.83576975, 2108.97501935,  542.19927004,  977.78246905,
       1194.72599137, 1089.96554234,  555.78416036, 1015.44496825,
        503.94937675, 1272.17000352,  579.79326545,  754.07498791,
       -108.12132561, 1314.5536642 , 1142.05723815, 1123.29941838,
        871.05075915, 1004.69899737, 1613.09739686, 1245.60544102,
       1083.08194805, 1211.86201634,  607.23917101, 1984.39478329,
       1903.94136615, 1289.86528221,  635.84253581, 1372.42883573,
       1042.27855197, 2109.37864488,  212.79365849, 1957.17294936,
        993.33223118, 1945.08878886,  736.31843084,  397.48040049,
        762.89545114, 1133.3367634 ,  659.43801876,  712.50284242,
       1310.40227415,  917.6789895 , 1126.53914372, 1227.17912

IMPORTANTE: APLICAR LO MISMO A ESTOS DATOS QUE HAYÁIS APLICADO A LOS DATOS DE ENTRENAMIENTO

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO BORRAR FILAS, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER 391 FILAS, SI O SI

**Entonces, si al cargar los datos de train usé `index_col=0` para que utilizara la primera columna del conjunto de datos como índice, ¿tendré que hacerlo también para el conjunto `test.csv`?**

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

In [16]:
X_pred = pd.read_csv("data/test2.csv", index_col=0)
X_pred.head()

Company  TypeName  Inches  Ram  OpSys  Weight  Memory1_1   
laptop_ID                                                             
750              9         0    14.0    8      0    1.25        256  \
726             14         0    14.0    4      0    1.40         32   
633             14         0    14.0    2      0    1.40         32   
363              1         3    15.6    8      0    2.65       1000   
319              4         4    11.6    2      0    1.10         32   

           Memory1_2  Memory2_1  Memory2_2  Resolution1  Gpu1  Gpu2  Cpu1   
laptop_ID                                                                   
750                1          0          4      2073600     1     3     1  \
726                0          0          4      2304000     1     3     1   
633                3          0          4      2073600     1     3     1   
363                0          0          4      2073600     2     2     1   
319                3          0          4      1049088     1     3     1   

           Cpu2  Cpu3  
laptop_ID              
750          12     2  
726          21    29  
633          21    15  
363          12     2  
319          15     6

### 3. AHORA puedo hacer la predicción que será lo que subirás a Kaggle. 

**¿Qué es lo que subirás a Kaggle?**

In [17]:
predictions_submit = model.predict(X_pred)
predictions_submit

array([1226.19776101,   28.35335923,  155.15443347, 1108.91859732,
        728.60547994,  483.48832199,  535.02008634, 1927.19313338,
       1814.89584674, 1314.5536642 , 1084.82894655, 1111.85381163,
       1313.36783648, 1018.88331038,  690.45177683, 1069.20563923,
       1233.83576975, 2108.97501935,  542.19927004,  977.78246905,
       1194.72599137, 1089.96554234,  555.78416036, 1015.44496825,
        503.94937675, 1272.17000352,  579.79326545,  754.07498791,
       -108.12132561, 1314.5536642 , 1142.05723815, 1123.29941838,
        871.05075915, 1004.69899737, 1613.09739686, 1245.60544102,
       1083.08194805, 1211.86201634,  607.23917101, 1984.39478329,
       1903.94136615, 1289.86528221,  635.84253581, 1372.42883573,
       1042.27855197, 2109.37864488,  212.79365849, 1957.17294936,
        993.33223118, 1945.08878886,  736.31843084,  397.48040049,
        762.89545114, 1133.3367634 ,  659.43801876,  712.50284242,
       1310.40227415,  917.6789895 , 1126.53914372, 1227.17912

**¡PERO! Para subir a Kaggle la predicción, ésta tendrá que tener una forma específica y no valdrá otra.**

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [19]:
sample = pd.read_csv("data/sample_submission.csv")

In [20]:
sample.head()

laptop_ID  Price_euros
0        750          500
1        726          500
2        633          500
3        363          650
4        319          650

In [22]:
sample.shape

(391, 2)

### 4. Mete tus predicciones en un dataframe. 

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [21]:
submission = pd.DataFrame({"laptop_ID": sample['laptop_ID'], "Price_euros": predictions_submit})

In [22]:
submission.head()

laptop_ID  Price_euros
0        750  1226.197761
1        726    28.353359
2        633   155.154433
3        363  1108.918597
4        319   728.605480

In [23]:
submission.shape

(391, 2)

### 5. Pásale el CHEQUEATOR para comprobar que efectivamente está listo para subir a Kaggle.

In [25]:
def chequeator(df_to_submit):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == sample.shape:
        if df_to_submit.columns.all() == sample.columns.all():
            if df_to_submit.laptop_ID.all() == sample.laptop_ID.all():
                print("You're ready to submit!")
                submission.to_csv("submission.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://i.kym-cdn.com/photos/images/facebook/000/747/556/27a.jpg", "gfg.png")     
                img = Image.open("gfg.png")
                img.show()   
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("\nMensaje secreto de Clara: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `diamonds_test.csv`. Lloro.")


In [26]:
chequeator(submission)

You're ready to submit!
